docs/apache-airflow/public-airflow-api.rst

```diff
@@ -0,0 +1,141 @@
+and developers who want to extend the functionality of the system.
+
+What kind of APIs are Public in Apache Airflow?
+===============================================
+
+Apache Airflow has a number of different public APIs that allow developers to interact with various
+aspects of the system. Some examples of the types of public APIs exposed as Python objects
+that are available in Apache Airflow include:
+
+* :doc:`DAG <concepts/dags>`_ (Directed Acyclic Graph) APIs, which allow developers to create, manage,
+  and access DAGs in Airflow.
+* :doc:`Task <concepts/tasks>`_ APIs, which provide access to information about individual tasks within
+  a DAG, such as their dependencies and execution status.
+* :doc:`Operator <concepts/operators>`_ APIs, which allows the developers to write their custom Operators.
+* :doc:`Decorators <howto/create-custom-decorator>`_ APIs, which allows the developers to write their
+  custom decorators to make it easier to write :doc:`TaskFlow <tutorial/taskflow>`_ DAGs.
+* :doc:`Secret Managers <security/secrets>`_ APIs, which allows the developers to write their custom
+  Secret Managers to safely access credentials and other secret configuration of their workflows.
+* :doc:`Connection management <concepts/connections>`_ APIs, which allow developers to manage
+  connections to external systems
+* :doc:`XCom <concepts/xcoms>`_, which allow developers to manage cross-task communication within Airflow.
+* :doc:`Variables <concepts/variables>`_, which allow developers to manage variables within Airflow.
+* :doc:`Executors <executor/index>`_, which allow developers to manage the execution of tasks within Airflow.
+* :doc:`Plugins <plugins>`_, which allow developers to extend internal Airflow capabilities - add new UI
+  pages, custom :doc:`TimeTables <concepts/timetable>`_, :doc:`Extra Links <howto/define_extra_link>`_,
+  :doc:`Listeners <listeners>`_ - all of which are considered public APIs.
+
+Also Airflow has a stable REST API that allows users to interact with Airflow via HTTP requests and a
+CLI that allows users to interact with Airflow via command line commands.
+
+Overall, the public APIs in Apache Airflow are designed to provide developers with a wide
+range of tools and capabilities for interacting with the system and extending its functionality.
+
+What is not part of the Public API of Apache Airflow?
+=====================================================
+
+A public API of Apache Airflow is a set of programming interfaces that are designed to be used
+by developers to access certain features and functionality of the system. As such, not everything in
+Apache Airflow is considered to be part of the public API.
+
+For example, the internal implementation details of Apache Airflow, such as the specific algorithms
+and data structures used to manage DAGs and tasks, are not considered to be part of the public API.
```

Comment:
This is done by GPT chat.

Reply:
I think we have (for now) a LOT of classes that should not be used but are not _underscored . Simply because it has never been created in the way to separte them. I am afraid for now at least what we can do is to be a bit vague. 

Later - when we introduce a formal list of classes - we can change it. But I want to  first document what we have rather than go into the rabbit hole of changing 100s of classes and methods to contain underscore. It's just pragmatic approach.

Reply:
> But I want to first document what we have rather than go into the rabbit hole of changing 100s of classes and methods to contain underscore.

Sorry for the confusion I was **not** saying that we should update those classes in this PR. I was just saying that we should add concrete verbiage here in this documentation that says any class/method/function that is "private" is not included in the public API.
It's an easy win and a useful tool we can use in the future (when we do start marking more things as private).

Reply:
I think we want to be MUCH stronger than that. Anything that is NOT mentioned in this docs should be considered Private. This is what I am adding in the new version of the docs. 

Currently we have almost no` _` classes in our codebase and we even have no convention agreed that we should do it. I'd say maybe at some point in the future we should have it, but even that I am not sure if we will do it, because this is pretty uncommon in the Python world and it simply looks ugly. So while maybe it is an easy win, I am not sure if we will ever go this direction and adding it at this stage assumes we will.

Simply speaking - let's not add a new convention in this doc that we neither have nor agreed to yet. My goal is to document current state at this stage and any future changes to conventions and agreements should be left for, well, future ....

Reply:
I just don't see the downside of including it. There will never be a time where we _want_ private methods/classes to be in the public API, so we can leverage it straight away as an actually useful mechanism without any harm. 

Most other things documented here are very debatable/vague (other than the DB section). I'm just looking for tools that can actually be applied in practice and I think private code is an easy win.

But, as always, I'm happy to disagree and approve if I'm out voted 👍